File: Step.cpp

#include "Step.hpp"

#include "Action.hpp"
#include "State.hpp"
#include "Cell.hpp"

void StepLeft::do_move() {
  action->next->cell = action->current->cell->left();
  action->do_next();
}

void StepLeft::print(std::ostream& os) const {
  os << "LEFT";
}

void StepRight::do_move() {
  action->next->cell = action->current->cell->right();
  action->do_next();
}

void StepRight::print(std::ostream& os) const {
  os << "RIGHT";
}

void StepNone::do_move() {
  action->next->cell = action->current->cell;
  action->do_next();
}

void StepNone::print(std::ostream& os) const {
  os << "NONE";
}

std::ostream& operator<<(std::ostream& os, const Step* step) {
  step->print(os);
  return os;
}

File: js/Step.js

class StepLeft {

  constructor() {
    this.action = null;
    Object.seal(this);
  }

  doMove() {
    this.action.next.cell = this.action.current.cell.left();
    this.action.doNext();
  }

  toString() {
    return "LEFT";
  }
}

class StepRight {

  constructor() {
    this.action = null;
    Object.seal(this);
  }

  doMove() {
    this.action.next.cell = this.action.current.cell.right();
    this.action.doNext();
  }

  toString() {
    return "RIGHT";
  }
}

class StepNone {

  constructor() {
    this.action = null;
    Object.seal(this);
  }

  doMove() {
    this.action.next.cell = this.action.current.cell;
    this.action.doNext();
  }

  toString() {
    return "NONE";
  }
}

module.exports = {StepLeft, StepRight, StepNone};